cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js

```diff
@@ -272,6 +276,54 @@ class HydratorPlusPlusTopPanelCtrl {
     };
   }
 
+  fetchMacros() {
+    console.log('calling fetch macros');
+    this.macrosMap = [];
+    let configStages = this.myHelpers.objectQuery(this.HydratorPlusPlusConfigStore.getDisplayConfig(), 'config', 'stages');
+
+    for (let i = 0; i < configStages.length; i++) {
+      let properties = this.myHelpers.objectQuery(configStages[i], 'plugin', 'properties');
+      for (let prop in properties) {
+        if (properties.hasOwnProperty(prop)) {
+          let propValue = properties[prop];
+          if (propValue.substring(0, 2) === '${' && propValue[propValue.length - 1] === '}') {
+            this.macrosMap[propValue.substring(2, propValue.length - 1)] = '';
+          }
+        }
+      }
+    }
```

Comment:
This `fetchMacros` function is actually different than the one in detail view, since we're getting the macro keys from two different places. However, the `syncPreferencesStoreWithMacros` function is the same in both controllers, so I can extract that out.